### PR TITLE
[Experimental] Add Doppler effect

### LINKED ISSUE #9172: Add Doppler effect
## Problem

With the introduction of the advtrains mod, entities were added that move at high speeds (up to 70 km/h), which already causes significant sound distortion due to doppler effect. 

## Solutions

Set the velocity correctly in the OpenAL sound object.

## Alternatives

Unrealistic sounds in advtrains.

## Additional context

I have tried implementing this. However, I have stumbled upon the issue of finding a CAO's velocity, which I do not know how to resolve.

### PR DESCRIPTION
**Pull request by gpcf**
Known bugs: This will not work for attached entities.

- Goal of the PR
Adding doppler effect, closes #9172.
- How does the PR work?
It passes the velocity of the object to OpenAL
## To do

This PR is a Work in Progress

## How to test

Download advtrains  and have trains passing by you to hear the doppler effect.

### CODE REVIEW

**Comment by rubenwardy** [comment]
I believe ClientActiveObject exists as an interface which makes testing things possible. Ideally, you shouldn't need to do the change to GenericCAO - but it appears to be *fine* anyway

Worth seeing what the unit tests come back with

**Comment by gpcf** [comment]
Tests failed, because the tests are broken....

**Comment by rubenwardy** [comment]
Is this clever enough to have the player's velocity affect it?

**Comment by rubenwardy** [comment]
Travis seems to like to throw a wobbly, I've restarted the build. The fail wasn't due to this PR

**Comment by gpcf** [comment]
I did not touch code affecting the listener, since I saw that minetest already sets the velocity for players. 

It seems to work, since I feel the sounds sound normal when mounted on the train.

**Comment by rubenwardy** [comment]
> It seems to work, since I feel the sounds sound normal when mounted on the train.

Ah, that makes sense

**Comment by SmallJoker** [comment]
Shouldn't this be a delta-velocity calculation? There's no distortion when you're riding the train.

File: src/client/content_cao.cpp [hunk 1]
```diff
@@ -412,6 +412,12 @@ const v3f GenericCAO::getPosition() const
 	return pos_translator.val_current;
 }
 
+const v3f GenericCAO::getVelocity() const
+{
+	/* TODO: handle attached entities */
+	return m_velocity;
```
Comment (SmallJoker) [inline src/client/content_cao.cpp]: Untested, but this should roughly look like this:

```C++
if (getParent())
	return getParent()->getVelocity();

if (m_is_localplayer)
	return m_env->getLocalPlayer()->getSpeed();

return m_velocity;
```

**Comment by gpcf** [comment]
The local player was afaik already handled by minetest previous to this PR, since MT sets the velocity of the listener. Then openAL computes doppler using the velocity of the listener and the sources.

**Comment by gpcf** [comment]
I wrote some code to handle attached entities, however, I still haven't got a mod to test it.

**Comment by SmallJoker** [comment]
@gpcf 
1) Install [advtrains](https://forum.minetest.net/viewtopic.php?f=11&t=14726) or carts (minetest_game)
2) Host a local game
3) Join with a second local player
4) Mute one player
5) Build a track (loop) and drive past each other real fast (tweak mod settings if it's too slow)

**Comment by gpcf** [comment]
Advtrains does not feature noisy entities attached to other entities. I have tested the doppler effect using advtrains and fast priv, running around a moving train.

**Comment by SmallJoker** [comment]
@gpcf Did you plan to continue working on this PR?

**Comment by gpcf** [comment]
yes, only that I need to write a small mod to test the attached entities stuff, and do more testing in general with non-advtrains entities.

**Comment by paramat** [comment]
Seems neglected.

**Pull request closed without merging**